ngraph/src/ngraph/node.hpp

```diff
@@ -489,6 +489,88 @@ namespace ngraph
+///        operations.
+///
+/// Use this macro as a public part of the class definition:
+///
+///     class MyOp : public Node
+///     {
+///         public:
+///             // Don't use Node as a parent for type_info, it doesn't have any value and
+///             prohibited
+///             NGRAPH_RTTI_DECLARATION("MyOp", 1);
+///
+///             ...
+///     };
+///
+///     class MyInheritedOp : public MyOp
+///     {
+///         public:
+///             NGRAPH_RTTI_DECLARATION("MyInheritedOp", 1, MyOp)
+///
+///             ...
+///     };
+///
+/// To complete type identification for a class, use NGRAPH_RTTI_DEFINITION.
+///
+#define NGRAPH_RTTI_DECLARATION                                                                    \
+    static const ::ngraph::Node::type_info_t type_info;                                            \
+    const ::ngraph::Node::type_info_t& get_type_info() const override;                             \
+    const ::ngraph::Node::type_info_t& get_type_info_static();
+
+#define _NGRAPH_RTTI_DEFINITION_COMMON(CLASS)                                                      \
+    const ::ngraph::Node::type_info_t& CLASS::get_type_info() const { return type_info; }          \
+    const ::ngraph::Node::type_info_t& CLASS::get_type_info_static() { return type_info; }
+#define _NGRAPH_RTTI_DEFINITION_WITH_PARENT(CLASS, TYPE_NAME, _VERSION_INDEX, PARENT_CLASS)        \
+    const ::ngraph::Node::type_info_t CLASS::type_info{                                            \
+        TYPE_NAME, _VERSION_INDEX, &PARENT_CLASS::type_info};                                      \
+    _NGRAPH_RTTI_DEFINITION_COMMON(CLASS)
+
+#define _NGRAPH_RTTI_DEFINITION_NO_PARENT(CLASS, TYPE_NAME, _VERSION_INDEX)                        \
+    const ::ngraph::Node::type_info_t CLASS::type_info{TYPE_NAME, _VERSION_INDEX};                 \
+    _NGRAPH_RTTI_DEFINITION_COMMON(CLASS)
+
+#define _NGRAPH_RTTI_DEFINITION_SELECTOR(_1, _2, _3, _4, NAME, ...) NAME
+
+/// Complementary to NGRAPH_RTTI_DECLARATION, this helper macro _defines_ items _declared_ by
+/// NGRAPH_RTTI_DECLARATION.
+/// Should be used outside the class definition scope in place where ODR is ensured.
+///
+/// \param CLASS is a C++ name of the class where corresponding NGRAPH_RTTI_DECLARATION was applied.
+///
+/// Examples (see corresponding declarations in NGRAPH_RTTI_DECLARATION description):
+///
+///     NGRAPH_RTTI_DEFINITION(MyOp)
+///     NGRAPH_RTTI_DEFINITION(MyInheritedOp)
+///
+/// For convenience, TYPE_NAME and CLASS name are recommended to be the same.
+///
+#define NGRAPH_RTTI_DEFINITION(...)                                                                \
+    EXPAND(_NGRAPH_RTTI_DEFINITION_SELECTOR(__VA_ARGS__,                                           \
+                                            _NGRAPH_RTTI_DEFINITION_WITH_PARENT,                   \
+                                            _NGRAPH_RTTI_DEFINITION_NO_PARENT)(__VA_ARGS__))
```

Comment:
all these macro look like implementation detail and even not used in this header. So, do we need to clutter up such common library headers? Maybe to move to a dedicated header?

Reply:
This stuff is required to be used in almost all (useful) classes derived from Node. It is closely related to Node (connected to get_type_info virtual function). So this is a tool to build the correct interface for Node-inherited classes. That's why the macros are defined within Node class definition.

In this commit we don't migrate all classes derived from Node to these macros, because it is too time consuming. Want to commit the base for the next move as a separate commit. Anyway, the new things introduced by these macro can co-exist with old RTTI definition and in this way we avoided mandatory bulk changes in all ngraph Node-derived classes, but bring a new value anyway.